README.md
# utero-archive

:warning: This is no longer maintained. :warning:

<img src="doc/utero-archive 2017-12-24.gif" alt="utero-archive 2017-12-24">

**Utero** is an operating system (for x86_64) written in [Crystal](https://crystal-lang.org/) *as much as possible*.

This is the *work in progress*.

## Requirements

* Crystal 0.23.0
  * llvm
  * Please see [All required libraries](https://github.com/crystal-lang/crystal/wiki/All-required-libraries)
* nasm
* grub-mkrescue
* xorriso
* qemu-system-x86_64
* mformat (included in mtools)
  * optional, See more details http://os.phil-opp.com/multiboot-kernel.html#creating-the-iso

## Usage
To make an ISO file and run on Qemu

``$ make run``

To make an ISO file

``$ make iso``

To make a binary file of the kernel

``$ make`` or ``$ make all``

To compile the OS in DEBUG mode and run on Qemu

``$ make debug``

**Note:** in DEBUG mode, logging uses the serial port `COM1` to write various
debugging information. `qemu` is configured to write the output of this serial
port to `/tmp/serial.log`.

To clean up

``$ make clean``

### Troubleshooting on Qemu
On a system that uses EFI boot, like dual boot macOS and Ubuntu(16.04) on MacBook Pro

Error on Qemu like this

```sh
Could not read from CDROM (code 0009)
```

The solution may be:  

```sh
$ sudo apt-get install grub-pc-bin
```

After the install `grub-pc-bin`, you will need to recreate the ISO file like this:

```sh
$ make clean
$ make # or make iso
$ make run
```

The following links saved my life:

https://github.com/Microsoft/WSL/issues/1043

http://intermezzos.github.io/book/appendix/troubleshooting.html#Could%20not%20read%20from%20CDROM%20%28code%200009%29

## Contributing

1. Fork it ( https://github.com/UteroOS/utero-archive/fork )
2. Clone it (including submodules)

  ```
  git clone --recursive <YOUR-FORKED-UTERO-URL>
  ```

  or

  ```
  git clone <YOUR-FORKED-UTERO-URL>
  git submodule update --init --recursive
  ```

3. Create your feature branch (git checkout -b my-new-feature)
4. Commit your changes (git commit -am 'Add some feature')
5. Push to the branch (git push origin my-new-feature)
6. Create a new Pull Request

## Contributors

- [noriyotcp](https://github.com/noriyotcp) Noriyo Akita - creator, maintainer, breeder of utero-archive

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any
additional terms or conditions.


CHANGELOG.md
# Change Log

## [v0.6.0](https://github.com/UteroOS/utero-archive/tree/v0.6.0) (2018-03-30)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.5.0...v0.6.0)

**Merged pull requests:**

- 201803 [\#58](https://github.com/UteroOS/utero-archive/pull/58) ([noriyotcp](https://github.com/noriyotcp))

## [v0.5.0](https://github.com/UteroOS/utero-archive/tree/v0.5.0) (2017-12-24)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.4.0...v0.5.0)

**Merged pull requests:**

- Reborn [\#57](https://github.com/UteroOS/utero-archive/pull/57) ([noriyotcp](https://github.com/noriyotcp))

## [v0.4.0](https://github.com/UteroOS/utero-archive/tree/v0.4.0) (2017-09-30)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.3.0...v0.4.0)

**Merged pull requests:**

- Implement core [\#56](https://github.com/UteroOS/utero-archive/pull/56) ([noriyotcp](https://github.com/noriyotcp))

## [v0.3.0](https://github.com/UteroOS/utero-archive/tree/v0.3.0) (2017-06-30)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.2.1...v0.3.0)

**Closed issues:**

- Drop repo's 'crystallang' topic [\#43](https://github.com/UteroOS/utero-archive/issues/43)
- Implement stdio functions [\#31](https://github.com/UteroOS/utero-archive/issues/31)
- Implementing functions for string [\#2](https://github.com/UteroOS/utero-archive/issues/2)

**Merged pull requests:**

- Implement tasks \(but it's incomplete\) [\#55](https://github.com/UteroOS/utero-archive/pull/55) ([noriyotcp](https://github.com/noriyotcp))
- Reconstruct directory structure [\#54](https://github.com/UteroOS/utero-archive/pull/54) ([noriyotcp](https://github.com/noriyotcp))
- long\_mode\_init: Implement switch\_context [\#53](https://github.com/UteroOS/utero-archive/pull/53) ([noriyotcp](https://github.com/noriyotcp))
- Refer to linker symbols [\#52](https://github.com/UteroOS/utero-archive/pull/52) ([noriyotcp](https://github.com/noriyotcp))
- Bump up Crystal version to 0.22.0 [\#51](https://github.com/UteroOS/utero-archive/pull/51) ([noriyotcp](https://github.com/noriyotcp))
- Display kernel info [\#50](https://github.com/UteroOS/utero-archive/pull/50) ([noriyotcp](https://github.com/noriyotcp))
- Expand stack size [\#49](https://github.com/UteroOS/utero-archive/pull/49) ([noriyotcp](https://github.com/noriyotcp))
- Call idt\_install and isrs\_install from Crystal [\#48](https://github.com/UteroOS/utero-archive/pull/48) ([noriyotcp](https://github.com/noriyotcp))
- Implement idt isrs [\#47](https://github.com/UteroOS/utero-archive/pull/47) ([noriyotcp](https://github.com/noriyotcp))
- Refactoring makefile [\#46](https://github.com/UteroOS/utero-archive/pull/46) ([noriyotcp](https://github.com/noriyotcp))
- Refactoring Scrn methods [\#45](https://github.com/UteroOS/utero-archive/pull/45) ([noriyotcp](https://github.com/noriyotcp))
- Connect C with Crystal by LibU [\#44](https://github.com/UteroOS/utero-archive/pull/44) ([noriyotcp](https://github.com/noriyotcp))

## [v0.2.1](https://github.com/UteroOS/utero-archive/tree/v0.2.1) (2017-04-01)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.2.0...v0.2.1)

**Merged pull requests:**

- Fix broken tests [\#42](https://github.com/UteroOS/utero-archive/pull/42) ([noriyotcp](https://github.com/noriyotcp))

## [v0.2.0](https://github.com/UteroOS/utero-archive/tree/v0.2.0) (2017-03-27)
[Full Changelog](https://github.com/UteroOS/utero-archive/compare/v0.1.1...v0.2.0)

**Implemented enhancements:**

- Make core strong [\#17](https://github.com/UteroOS/utero-archive/issues/17)
- Enhance the terminal [\#15](https://github.com/UteroOS/utero-archive/issues/15)

**Closed issues:**

- Move io.cr \(inb & outb\) [\#39](https://github.com/UteroOS/utero-archive/issues/39)
- Restructure src/lib\_cr according to the directory structure of Crystal's src/lib\_c [\#38](https://github.com/UteroOS/utero-archive/issues/38)
- Introducing musl [\#34](https://github.com/UteroOS/utero-archive/issues/34)
- Implement inb function [\#32](https://github.com/UteroOS/utero-archive/issues/32)
- Review assembly codes [\#29](https://github.com/UteroOS/utero-archive/issues/29)
- Bump up Crystal version to 0.21.0 [\#25](https://github.com/UteroOS/utero-archive/issues/25)
- cooperation [\#21](https://github.com/UteroOS/utero-archive/issues/21)
- Implementing ctype functions \(more\) [\#14](https://github.com/UteroOS/utero-archive/issues/14)
- Implementing ctype functions [\#12](https://github.com/UteroOS/utero-archive/issues/12)
- Fix testing [\#4](https://github.com/UteroOS/utero-archive/issues/4)

**Merged pull requests:**

- Restructure for libcr \#38 [\#41](https://github.com/UteroOS/utero-archive/pull/41) ([noriyotcp](https://github.com/noriyotcp))
- Move io.cr \(inb & outb\) to src/lib\_cr/bits [\#40](https://github.com/UteroOS/utero-archive/pull/40) ([noriyotcp](https://github.com/noriyotcp))
- Change target option to x86\_64-unknown-linux-musl [\#37](https://github.com/UteroOS/utero-archive/pull/37) ([noriyotcp](https://github.com/noriyotcp))
- \[Makefile\] Make directories for object files [\#36](https://github.com/UteroOS/utero-archive/pull/36) ([noriyotcp](https://github.com/noriyotcp))
- Introducing musl \#34 [\#35](https://github.com/UteroOS/utero-archive/pull/35) ([noriyotcp](https://github.com/noriyotcp))
- Feature/\#32 implement inb [\#33](https://github.com/UteroOS/utero-archive/pull/33) ([noriyotcp](https://github.com/noriyotcp))
- Gdt without data segment [\#30](https://github.com/UteroOS/utero-archive/pull/30) ([noriyotcp](https://github.com/noriyotcp))
- \[scrn\] To avoid a bug using an explicit type annotation [\#28](https://github.com/UteroOS/utero-archive/pull/28) ([noriyotcp](https://github.com/noriyotcp))
- Re: Feature/\#17 make core strong [\#27](https://github.com/UteroOS/utero-archive/pull/27) ([noriyotcp](https://github.com/noriyotcp))

src/core/isr.c
#include <core/isr.h>
#include <core/ports.h>
#include <drivers/screen.h>
#include <kernel/panic.h>
#include <libk/printfk.h>
#include <musl/include/stdio.h>
#include <musl/include/stdlib.h>

#define NB_REGISTERS_PUSHED_BEFORE_CALL 15

stack_t* get_stack(uint64_t id, uint64_t stack);
void breakpoint_handler(stack_t *stack);

isr_t interrupt_handlers[256];
const char* exception_messages[] = {
    "Division By Zero",
    "Debug",
    "Non Maskable Interrupt",
    "Breakpoint",
    "Into Detected Overflow",
    "Out of Bounds",
    "Invalid Opcode",
    "No Coprocessor",

    "Double Fault",
    "Coprocessor Segment Overrun",
    "Bad TSS",
    "Segment Not Present",
    "Stack Fault",
    "General Protection Fault",
    "Page Fault",
    "Unknown Interrupt",

    "Coprocessor Fault",
    "Alignment Check",
    "Machine Check",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",

    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved",
    "Reserved"
};

void isr_init()
{
    // start initialization
    port_byte_out(PIC1, 0x11);
    port_byte_out(PIC2, 0x11);

    // set IRQ base numbers for each PIC
    port_byte_out(PIC1_DATA, IRQ_BASE);
    port_byte_out(PIC2_DATA, IRQ_BASE + 8);

    // use IRQ number 2 to relay IRQs from the slave PIC
    port_byte_out(PIC1_DATA, 0x04);
    port_byte_out(PIC2_DATA, 0x02);

    // finish initialization
    port_byte_out(PIC1_DATA, 0x01);
    port_byte_out(PIC2_DATA, 0x01);

    // mask all IRQs
    port_byte_out(PIC1_DATA, 0x00);
    port_byte_out(PIC2_DATA, 0x00);

    set_idt_gate(0, (uint64_t) isr0);
    set_idt_gate(1, (uint64_t) isr1);
    set_idt_gate(2, (uint64_t) isr2);
    set_idt_gate(3, (uint64_t) isr3);
    set_idt_gate(4, (uint64_t) isr4);
    set_idt_gate(5, (uint64_t) isr5);
    set_idt_gate(6, (uint64_t) isr6);
    set_idt_gate(7, (uint64_t) isr7);
    set_idt_gate(8, (uint64_t) isr8);
    set_idt_gate(9, (uint64_t) isr9);
    set_idt_gate(10, (uint64_t) isr10);
    set_idt_gate(11, (uint64_t) isr11);
    set_idt_gate(12, (uint64_t) isr12);
    set_idt_gate(13, (uint64_t) isr13);
    set_idt_gate(14, (uint64_t) isr14);
    set_idt_gate(15, (uint64_t) isr15);
    set_idt_gate(16, (uint64_t) isr16);
    set_idt_gate(17, (uint64_t) isr17);
    set_idt_gate(18, (uint64_t) isr18);
    set_idt_gate(19, (uint64_t) isr19);
    set_idt_gate(20, (uint64_t) isr20);
    set_idt_gate(21, (uint64_t) isr21);
    set_idt_gate(22, (uint64_t) isr22);
    set_idt_gate(23, (uint64_t) isr23);
    set_idt_gate(24, (uint64_t) isr24);
    set_idt_gate(25, (uint64_t) isr25);
    set_idt_gate(26, (uint64_t) isr26);
    set_idt_gate(27, (uint64_t) isr27);
    set_idt_gate(28, (uint64_t) isr28);
    set_idt_gate(29, (uint64_t) isr29);
    set_idt_gate(30, (uint64_t) isr30);
    set_idt_gate(31, (uint64_t) isr31);

    set_idt_gate(IRQ0, (uint64_t) irq0);
    set_idt_gate(IRQ1, (uint64_t) irq1);
    set_idt_gate(IRQ2, (uint64_t) irq2);
    set_idt_gate(IRQ3, (uint64_t) irq3);
    set_idt_gate(IRQ4, (uint64_t) irq4);

    // handlers for isr exceptions
    register_interrupt_handler(EXCEPTION_BP, breakpoint_handler);

    set_idt();
}

void irq_init()
{
    __asm__("sti");
}

void irq_disable()
{
    __asm__("cli");
}

void isr_handler(uint64_t id, uint64_t stack_addr)
{
    stack_t *stack = get_stack(id, stack_addr);

    if (interrupt_handlers[id] != 0) {
        isr_t handler = interrupt_handlers[id];
        handler(get_stack(id, stack_addr));

        return;
    }

    PANIC(
        "Received interrupt: %d - %s\n\n"
        "  instruction_pointer = 0x%X\n"
        "  code_segment        = 0x%X\n"
        "  cpu_flags           = 0x%X\n"
        "  stack_pointer       = 0x%X\n"
        "  stack_segment       = 0x%X",
        id, exception_messages[id],
        stack->instruction_pointer,
        stack->code_segment,
        stack->cpu_flags,
        stack->stack_pointer,
        stack->stack_segment
    );
}

void irq_handler(uint64_t id, uint64_t stack_addr)
{
    if (id >= 40) {
        port_byte_out(PIC2, PIC_EOI);
    }

    port_byte_out(PIC1, PIC_EOI);

    if (interrupt_handlers[id] != 0) {
        isr_t handler = interrupt_handlers[id];
        handler(get_stack(id, stack_addr));
    }
}

void register_interrupt_handler(uint64_t id, isr_t handler)
{
    interrupt_handlers[id] = handler;
}

stack_t* get_stack(uint64_t id, uint64_t stack_addr)
{
    // https://github.com/0xAX/linux-insides/blob/master/interrupts/interrupts-3.md
    //
    //     +------------+
    // +40 | %SS        |
    // +32 | %RSP       |
    // +24 | %RFLAGS    |
    // +16 | %CS        |
    //  +8 | %RIP       |
    //   0 | ERROR CODE | <-- %RSP
    //     +------------+
    //
    switch (id) {
        case 8:
        case 10:
        case 11:
        case 12:
        case 13:
        case 14:
        case 17:
            // skip error code, so that we always get the same stack_t
            stack_addr += sizeof(uint64_t);
            break;
    }

    return (stack_t*) (stack_addr + (NB_REGISTERS_PUSHED_BEFORE_CALL * sizeof(uint64_t)));
}

void breakpoint_handler(stack_t *stack)
{
    printfk(
        "Exception: BREAKPOINT\n"
        "  instruction_pointer = 0x%X\n"
        "  code_segment        = 0x%X\n"
        "  cpu_flags           = 0x%X\n"
        "  stack_pointer       = 0x%X\n"
        "  stack_segment       = 0x%X\n",
        stack->instruction_pointer,
        stack->code_segment,
        stack->cpu_flags,
        stack->stack_pointer,
        stack->stack_segment
    );
}
